pkg/envoy/secretsync.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package envoy
+
+import (
+	"context"
+	"fmt"
+
+	envoy_config_core_v3 "github.com/cilium/proxy/go/envoy/config/core/v3"
+	envoy_entensions_tls_v3 "github.com/cilium/proxy/go/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	tlsCrtAttribute = "tls.crt"
+	tlsKeyAttribute = "tls.key"
+
+	// Key for CA certificate is fixed as 'ca.crt' even though is not as "standard"
+	// as 'tls.crt' and 'tls.key' are via k8s tls secret type.
+	caCrtAttribute = "ca.crt"
+)
+
+// secretSyncer is responsible to sync K8s TLS Secrets in pre-defined namespaces
+// via xDS SDS to Envoy.
+type secretSyncer struct {
+	logger         logrus.FieldLogger
+	envoyXdsServer XDSServer
+
+	mutex   lock.Mutex
+	secrets map[resource.Key]*slim_corev1.Secret
+}
+
+func newSecretSyncer(logger logrus.FieldLogger, envoyXdsServer XDSServer) *secretSyncer {
+	return &secretSyncer{
+		logger:         logger,
+		envoyXdsServer: envoyXdsServer,
+		secrets:        map[resource.Key]*slim_corev1.Secret{},
+	}
+}
+
+func (r *secretSyncer) handleSecretEvent(ctx context.Context, event resource.Event[*slim_corev1.Secret]) error {
+	scopedLogger := r.logger.
+		WithField(logfields.K8sNamespace, event.Key.Namespace).
+		WithField("name", event.Key.Name)
+
+	var err error
+
+	switch event.Kind {
+	case resource.Upsert:
+		scopedLogger.Debug("Received Secret upsert event")
+		err = r.handleUpsert(ctx, event.Key, event.Object)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle Secret upsert")
+			err = fmt.Errorf("failed to handle CEC upsert: %w", err)
+		}
+	case resource.Delete:
+		scopedLogger.Debug("Received Secret delete event")
+		err = r.handleDelete(ctx, event.Key)
+		if err != nil {
+			scopedLogger.WithError(err).Error("failed to handle Secret delete")
+			err = fmt.Errorf("failed to handle Secret delete: %w", err)
+		}
+	}
+
+	event.Done(err)
+
+	return nil
+}
+
+func (r *secretSyncer) handleUpsert(ctx context.Context, key resource.Key, secret *slim_corev1.Secret) error {
+	r.mutex.Lock()
```

Comment:
Might be holding the lock for a while. Should we provide more granularity?

Reply:
good point. on second though i realized that we might not need this mutex (and its related map of "old" / applied secrets) at all. This was only used to provide it to the existing functionality that uses the old secret to compare it with the new one. 

But i don't think that this is necessary as the `resource.Resource` should only fire events in case of changes.

This way we could get rid of this complexity.

(The same approach is required in other cases of xDS (e.g. CiliumEnvoyConfig). But only to know which xDS resources we created in a previous reconciliation. We have to keep track of them to ensure we delete them if they no longer exist in the modified CiliumEnvoyConfig. But in case of the Secret its a 1:1 mapping between the k8s secret and the (single) xDS secret we map it to. And the namespace and name of the k8s secret can't change over its lifetime.)

@sayboras wdyt? i already changed it - but i can revert it if there are good reasons.

Reply:
IIRC, we used to have an issue in which any change in original secret is not synced to the new namespace if there are multiple changes in short duration. Another case is that  new reconciliation was not triggered if there is value changed for TLS secret used in either Ingress or Gateway API.

With the new controller-runtime, I think it will be fine to remove this mutex. One case we need to check is probably same secret is used in multiple Ingress/Gateway objects.

Reply:
Thanks for the info @sayboras 

> IIRC, we used to have an issue in which any change in original secret is not synced to the new namespace if there are multiple changes in short duration.

if this re-occurs - we should be able to limit this via `stream.Debounce` (https://github.com/cilium/cilium/blob/main/pkg/stream/operators.go#L194-L199)

> With the new controller-runtime, I think it will be fine to remove this mutex. 

it's not that the `mutex` was there before. i introduced it with this PR, because we no longer get the "old state". but then realized that it might not be required at all.

> Another case is that new reconciliation was not triggered if there is value changed for TLS secret used in either Ingress or Gateway API.

The implementation should watch all 3 potential secret namespaces (pure cec, ingress & gateway api). And respects whether they use different or the same namespace.